src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java

```diff
@@ -0,0 +1,384 @@
+        }
+        return networkLayer;
+    }
+
+    /**
+     * Creates an {@link SSLEngine} for the specified {@link Socket}. The {@link SSLContext} may have retained session
+     * state and thus enable a shorter handshake.
+     *
+     * @param socket the socket.
+     * @return the {@link SSLEngine}.                                             s
+     */
+    private SSLEngine createSSLEngine(Socket socket) {
+        SocketAddress remoteSocketAddress = socket.getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress) {
+            InetSocketAddress remoteInetAddress = (InetSocketAddress) remoteSocketAddress;
+            return context.createSSLEngine(remoteInetAddress.getHostName(), remoteInetAddress.getPort());
+        } else {
+            return context.createSSLEngine();
+        }
+    }
+
+    /**
+     * Handles the state transitions for a connection.
+     */
+    private class Handler extends Channel.Listener implements SSLEngineFilterLayer.Listener,
+            ConnectionHeadersFilterLayer.Listener, ChannelApplicationLayer.Listener, ProtocolStack.Listener,
+            ChannelApplicationLayer.ChannelDecorator {
+        /**
+         * The event.
+         */
+        @Nonnull
+        private final Jnlp4ConnectionState event;
+        /**
+         * The client database used when {@link JnlpProtocol4Handler#handle(Socket, Map, List)} is handling a client.
+         */
+        private JnlpClientDatabase clientDatabase;
+
+        /**
+         * {@code true} when invoked from {@link JnlpProtocol4Handler#connect(Socket, Map, List)}, {@code false} when
+         * invoked from {@link JnlpProtocol4Handler#handle(Socket, Map, List)}.
+         */
+        private boolean client;
+
+        /**
+         * Internal constructor for {@link JnlpProtocol4Handler#connect(Socket, Map, List)}.
+         *
+         * @param event the event.
+         */
+        Handler(Jnlp4ConnectionState event) {
+            this.event = event;
+            this.client = true;
+        }
+
+        /**
+         * Internal constructor for {@link JnlpProtocol4Handler#handle(Socket, Map, List)}.
+         *
+         * @param event          the event.
+         * @param clientDatabase the client database.
+         */
+        Handler(@Nonnull Jnlp4ConnectionState event, JnlpClientDatabase clientDatabase) {
+            this.event = event;
+            this.clientDatabase = clientDatabase;
+            this.client = false;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void onHandshakeCompleted(SSLSession session) throws ConnectionRefusalException {
+            X509Certificate remoteCertificate;
+            try {
+                remoteCertificate = (X509Certificate) session.getPeerCertificates()[0];
+            } catch (ClassCastException e) {
+                // shouldn't happen as handshake is completed, but just in case
```

Comment:
May also happen if remote classloading rejects a class or any subclass of X509Certificate